File: app/src/flink/PortfolioRecalculator.java
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer;
import org.apache.flink.util.Collector;
import org.apache.flink.api.java.utils.ParameterTool;

import org.json.JSONObject;

import java.util.Properties;

public class PortfolioRecalculator {

    public static void main(String[] args) throws Exception {
        // Set up the execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // Read parameters for Kafka configuration
        final ParameterTool params = ParameterTool.fromArgs(args);
        String kafkaBootstrapServers = params.get("kafka.bootstrap.servers", "my-cluster-kafka-bootstrap:9092");
        String inputTopic = params.get("input.topic", "user-transactions");
        String outputTopic = params.get("output.topic", "user-portfolio");

        // Set up Kafka consumer
        Properties consumerProps = new Properties();
        consumerProps.setProperty("bootstrap.servers", kafkaBootstrapServers);
        consumerProps.setProperty("group.id", "portfolio-recalculator");
        FlinkKafkaConsumer<String> kafkaConsumer = new FlinkKafkaConsumer<>(inputTopic, new SimpleStringSchema(), consumerProps);

        // Set up Kafka producer
        Properties producerProps = new Properties();
        producerProps.setProperty("bootstrap.servers", kafkaBootstrapServers);
        FlinkKafkaProducer<String> kafkaProducer = new FlinkKafkaProducer<>(outputTopic, new SimpleStringSchema(), producerProps);

        // Consume transactions from Kafka
        DataStream<String> transactions = env.addSource(kafkaConsumer);

        // Process transactions to recalculate the portfolio
        DataStream<String> recalculatedPortfolio = transactions
            .flatMap((String transaction, Collector<Tuple3<String, String, Integer>> out) -> {
                // Parse JSON transaction
                JSONObject json = new JSONObject(transaction);
                String user = json.getString("user");
                String stock = json.getString("stock");
                String transactionType = json.getString("transaction_type");
                int quantity = json.getInt("quantity");

                // Emit (user, stock, quantity change)
                int delta = transactionType.equalsIgnoreCase("buy") ? quantity : -quantity;
                out.collect(new Tuple3<>(user, stock, delta));
            })
            .returns(TypeInformation.of(Tuple3.class))
            .keyBy(t -> t.f0 + "-" + t.f1) // Group by user and stock
            .sum(2) // Aggregate quantity changes
            .map(tuple -> {
                // Convert aggregated data to JSON
                JSONObject json = new JSONObject();
                json.put("user", tuple.f0);
                json.put("stock", tuple.f1);
                json.put("quantity", tuple.f2);
                return json.toString();
            });

        // Write recalculated portfolio to Kafka
        recalculatedPortfolio.addSink(kafkaProducer);

        // Execute the Flink job
        env.execute("Portfolio Recalculator");
    }
}

File: app/src/flink/portfolio_recalculator.py
import os
import json
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.datastream.connectors import FlinkKafkaConsumer, FlinkKafkaProducer
from pyflink.common.serialization import SimpleStringSchema


def recalculate_portfolio(transaction, current_prices, current_portfolios):
    """Update portfolio based on a transaction and market prices."""
    user = transaction['user']
    stock = transaction['stock']
    quantity = transaction['quantity'] if transaction['transaction_type'] == "buy" else -transaction['quantity']

    # Update portfolio quantities
    if user not in current_portfolios:
        current_portfolios[user] = {}
    current_portfolios[user][stock] = current_portfolios[user].get(stock, 0) + quantity

    # Calculate portfolio value using market prices
    portfolio_value = sum(
        quantity * current_prices.get(stock, 0)
        for stock, quantity in current_portfolios[user].items()
    )

    return {
        "user": user,
        "portfolio": current_portfolios[user],
        "total_value": portfolio_value
    }


def process_streams(transaction_json, market_json, portfolios_state, market_prices):
    """Main processing logic for transactions and market data."""
    try:
        # Update market prices
        if market_json:
            stock, price = market_json.split(":")
            market_prices[stock] = float(price)
            return None  # Market data doesn't produce an output directly

        # Process transactions
        if transaction_json:
            transaction = json.loads(transaction_json)
            updated_portfolio = recalculate_portfolio(transaction, market_prices, portfolios_state)
            return json.dumps(updated_portfolio)

    except Exception as e:
        print(f"Error processing stream: {e}")
        return None


def main():
    # Environment variables
    kafka_bootstrap_servers = os.getenv("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
    transactions_topic = os.getenv("TRANSACTIONS_TOPIC", "user-transactions")
    market_data_topic = os.getenv("MARKET_DATA_TOPIC", "market-data")
    output_topic = os.getenv("OUTPUT_TOPIC", "portfolio-updates")
    group_id = os.getenv("GROUP_ID", "portfolio-recalculator")

    # Set up the Flink execution environment
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(4)

    # Kafka consumers
    transaction_source = FlinkKafkaConsumer(
        topics=transactions_topic,
        deserialization_schema=SimpleStringSchema(),
        properties={
            'bootstrap.servers': kafka_bootstrap_servers,
            'group.id': group_id
        }
    )
    market_source = FlinkKafkaConsumer(
        topics=market_data_topic,
        deserialization_schema=SimpleStringSchema(),
        properties={
            'bootstrap.servers': kafka_bootstrap_servers,
            'group.id': group_id
        }
    )

    # Kafka producer
    kafka_sink = FlinkKafkaProducer(
        topic=output_topic,
        serialization_schema=SimpleStringSchema(),
        producer_config={
            'bootstrap.servers': kafka_bootstrap_servers
        }
    )

    # In-memory state
    current_portfolios = {}
    market_prices = {}

    # Stream processing pipeline
    transaction_stream = env.from_source(transaction_source, "transaction-source", None)
    market_stream = env.from_source(market_source, "market-source", None)

    transaction_stream.union(market_stream) \
        .map(lambda data: process_streams(data if "transaction_type" in data else None,
                                          data if ":" in data else None,
                                          current_portfolios, market_prices)) \
        .filter(lambda x: x is not None) \
        .add_sink(kafka_sink)

    # Execute the Flink job
    env.execute("Portfolio Recalculator with Market Data")


if __name__ == "__main__":
    main()
